core/src/main/java/org/jobrunr/server/carbonaware/CarbonAwareJobManager.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    private void scheduleJobAtOptimalTime(Job job, CarbonAwareAwaitingState carbonAwareAwaitingState) {
+        Instant lowestCarbonIntensityInstant = carbonIntensityForecast.lowestCarbonIntensityInstant(carbonAwareAwaitingState.getFrom(), carbonAwareAwaitingState.getTo());
+        carbonAwareAwaitingState.moveToNextState(job, lowestCarbonIntensityInstant);
+    }
+
+    private void scheduleJobAtPreferredInstant(Job job, CarbonAwareAwaitingState carbonAwareAwaitingState, String reason) {
+        Instant scheduleAt = isNull(carbonAwareAwaitingState.getPreferredInstant()) ? carbonAwareAwaitingState.getFrom() : carbonAwareAwaitingState.getPreferredInstant();
+        carbonAwareAwaitingState.moveToNextState(job, scheduleAt, reason);
+    }
+
+    private boolean isDeadlinePassed(CarbonAwareAwaitingState carbonAwareAwaitingState) {
+        return Instant.now().isAfter(carbonAwareAwaitingState.getTo());
+    }
+
+    private CarbonIntensityApiClient createCarbonIntensityApiClient(CarbonAwareConfigurationReader carbonAwareConfiguration, JsonMapper jsonMapper) {
+        return new CarbonIntensityApiClient(carbonAwareConfiguration, jsonMapper);
+    }
+
+    @VisibleFor("testing")
+    void updateCarbonIntensityForecast() {
+        CarbonIntensityForecast carbonIntensityForecast = carbonIntensityApiClient.fetchCarbonIntensityForecast();
+        if (carbonIntensityForecast.hasNoForecast() && !carbonIntensityForecast.hasError()) {
+            LOGGER.warn("No new carbon intensity forecast available. Keeping the old forecast.");
+            return;
+        }
+        this.carbonIntensityForecast = carbonIntensityForecast;
+    }
+
+    private void updateNextRefreshTime() {
+        Instant nextForecastAvailableAt = carbonIntensityForecast.getNextForecastAvailableAt();
+        Instant defaultDailyRefreshTime = getDefaultDailyRefreshTime();
+
+        Instant proposedNextRefreshTime = nonNull(nextForecastAvailableAt) ? nextForecastAvailableAt : defaultDailyRefreshTime;
+
+        if (proposedNextRefreshTime.isBefore(Instant.now())) {
+            nextRefreshTime = defaultDailyRefreshTime.plus(1, DAYS).plusSeconds(randomRefreshTimeOffset);
+        } else {
+            nextRefreshTime = proposedNextRefreshTime.plusSeconds(randomRefreshTimeOffset);
+        }
+    }
+
+    @VisibleFor("testing")
+    Instant getNextRefreshTime() {
+        return nextRefreshTime;
+    }
+
+    @VisibleFor("testing")
+    ZoneId getTimeZone() {
+        return nonNull(carbonIntensityForecast.getTimezone()) ? ZoneId.of(carbonIntensityForecast.getTimezone()) : ZoneId.systemDefault();
+    }
+
+    @VisibleFor("testing")
+    Instant getDefaultDailyRefreshTime() {
+        return ZonedDateTime.now(getTimeZone())
+                .truncatedTo(HOURS)
+                .withHour(DEFAULT_REFRESH_TIME).toInstant();
+    }
```

Comment:
If we put them visibleForTesting, is it then the right place to put them?

Reply:
I don't think they'll be useful anywhere else. These are really internal to the CarbonAwareJobManager. The only reason they are made visible is to be able to confirm that internal state is changed when performing some actions.

We can probably do without tests for `getTimeZone` and `getDefaultDailyRefreshTime` as they are included in the computation of `nextRefreshTime`.  The only way, I can think of, to check if `nextRefreshTime` is updated, is using `updateCarbonIntensityForecast` and mock time at different instants. But `getNextRefreshTime` could probably be public.